### PR TITLE
RFC: Fine-grained Timeout Configuration

### PR DESCRIPTION
**Pull request by Velfi**
## Motivation and Context

Provides context on how timeout configuration should work

## Description

Provides context on how timeout configuration should work

## Testing

N/A

## Checklist
- [x] I have updated `CHANGELOG.md`
- [x] I have updated `aws/SDK_CHANGELOG.md` if applicable

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+- **Service-specific Config**: A code-generated `Config` that has methods for setting service-specific configuration.
+  Each `Config` is defined in the `config` module of its parent service. For example, the S3-specific config struct
+  is `use`able from `aws_sdk_s3::config::Config` and re-exported as `aws_sdk_s3::Config`. In this case, "service" refers to an AWS offering like S3.
+
+### HTTP stack terms
+
+- **Service**: A trait defined in the [`tower-service` crate][tower_service::Service]. The lowest level of abstraction we deal with when making HTTP requests. Services act directly on data to transform and modify that data. A Service is what eventually turns a request into a response.
+- **Layer**: Layers are a higher-order abstraction over services that is used to compose multiple services together, creating a new service from that combination. Nothing prevents us from manually wrapping services withing services, but Layers allow us to do it in a maintainable, flexible, and generic manner. Layers don't directly act on data but produce a new compose service that does. Layers can be thought of as middleware.
```
Comment (rcoh) [inline design/src/rfcs/rfc0007_timeout_config.md:35]:
```suggestion
- **Layer**: Layers are a higher-order abstraction over services that is used to compose multiple services together, creating a new service from that combination. Nothing prevents us from manually wrapping services withing services, but Layers allow us to do it in a maintainable, flexible, and generic manner. Layers don't directly act on data but instead can wrap an existing service with additional functionality, creating a new service.. Layers can be thought of as middleware.
```

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+- **Middleware**: a term with several meanings,
+  - Generically speaking, middlewares are similar to Services and Layers in that they modify requests and responses.
+  - In the SDK, Middlewares have the specific meaning of being the only kind of struct that contains a `DispatchService`.
```
Comment (rcoh) [inline design/src/rfcs/rfc0007_timeout_config.md:38]: this is slightly more subtle. A middleware is a layer which can only wrap a service whose input type is `smithy_http::operation::Request` and whose output type is `operation::Response` (the dispatch service has that shape).

I might say:
- In the SDK, "Middleware" refers to a layer that can be wrapped around a `DispatchService`. In practice, this means that the resulting `Service` (and the inner service) must meet the bound `T: where T: Service<operation::Request, Response=operation::Response, Error=SendOperationError>`.

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  - Examples of Middleware include [MapRequest], [AsyncMapRequest], and [ParseResponse]. **TODO** is this true or am I getting things mixed up?
```
Comment (rcoh) [inline design/src/rfcs/rfc0007_timeout_config.md:39]: I would probably add `AwsMiddleware`, the combination of all middlewares required to assemble an SDK

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+- **DispatchService**: Connects Operation driven middleware to an HTTP implementation. It will also wrap the error type in OperationError to enable operation middleware reporting specific errors.
+  - **TODO** This definition is taken directly from a code comment. It makes it sound like this is the first and last layer that gets hit when making a request; Is that actually the case?
```
Comment (rcoh) [inline design/src/rfcs/rfc0007_timeout_config.md:41]: it is the "most inside" layer of the onion, if you will

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+- **Connector**: a term with several meanings,
+  - (esp. `DynConnector`) generally refers to a service that handles HTTP. HTTP is the lowest level we handle
+    - Dyn means that we're erasing the inner type
+    - **TODO** is this then a Service that we've erased the type of?
```
Comment (rcoh) [inline design/src/rfcs/rfc0007_timeout_config.md:45]: yep! A service like `T where T: tower::Service<http::Request, Response=http::Response>`

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+### Timeout terms
+
+- **Connect Timeout**: A limit on the amount of time after making an initial connect attempt on a socket to complete the
+  connect-handshake.
+    - _TODO: the runtime is based on Hyper which reuses connection and doesn't currently have a way of guaranteeing that
+      a fresh connection will be use for a given request._
+- **TLS Negotiation Timeout**: A limit on the amount of time a TLS handshake takes from when the CLIENT HELLO message is
+  sent to the time the client and server have fully negotiated ciphers and exchanged keys.
+    - _QUESTION: Our HTTP client supports multiple TLS implementations. Is there any reason that that would make
```
Comment (rcoh) [inline design/src/rfcs/rfc0007_timeout_config.md:58]: yeah, we probably need library support to enable this

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+      implementing this feature difficult?_
+- **Time to First Byte Timeout**: A limit on the amount of time an application takes to attempt to read the first byte over
```
Comment (rcoh) [inline design/src/rfcs/rfc0007_timeout_config.md:60]: I thing a `write timeout` refers to the send-side equivalent if time to first byte

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  an established, open connection after write request.
+  - _QUESTION: Is this the same as a "write timeout"? If yes, should we refer to it by that name? Also, if yes, should we not also support "read timeouts" for completeness' sake?_
+- **HTTP Request Timeout For A Single Request**: A limit on the amount of time it takes for the first byte to be sent over
+  an established, open connection and when the last byte is received from the service.
+- **HTTP Request Timeout For Multiple Attempts**: This timeout acts like the previous timeout but constrains the total time
+  it takes to make a request plus any retries.
+    - _NOTE: In a way, this is already possible in that users are free to race requests against timer futures with
```
Comment (rcoh) [inline design/src/rfcs/rfc0007_timeout_config.md:67]: yeah! we _may_ want to provide this out of the box to help ensure users don't ship code that will wait forever

**Review by jdisanti — COMMENTED** [review]
Great start! I'm excited to see how the Behind the Scenes section evolves.

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+=======================================
+
+> Status: RFC
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+It is not currently possible for users of the SDK to configure a client's maximum number of retry attempts. This RFC
+establishes a method for users to set the number of retries to attempt when calling a service and would allow users to
+disable retries entirely. This RFC would introduce breaking changes to the `retry` module of the `aws-smithy-client`
+crate.
```
[the remark this answers is not in the document]
Comment (Velfi) [reply design/src/rfcs/rfc0007_timeout_config.md:11]: oops

File: design/src/rfcs/rfc0007_timeout_config.md [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+- **Service-specific Config**: A code-generated `Config` that has methods for setting service-specific configuration.
+  Each `Config` is defined in the `config` module of its parent service. For example, the S3-specific config struct
+  is `use`able from `aws_sdk_s3::config::Config` and re-exported as `aws_sdk_s3::Config`. In this case, "service" refers to an AWS offering like S3.
+
+### HTTP stack terms
+
+- **Service**: A trait defined in the [`tower-service` crate][tower_service::Service]. The lowest level of abstraction we deal with when making HTTP requests. Services act directly on data to transform and modify that data. A Service is what eventually turns a request into a response.
+- **Layer**: Layers are a higher-order abstraction over services that is used to compose multiple services together, creating a new service from that combination. Nothing prevents us from manually wrapping services withing services, but Layers allow us to do it in a maintainable, flexible, and generic manner. Layers don't directly act on data but instead can wrap an existing service with additional functionality, creating a new service.. Layers can be thought of as middleware.
```
[the remark this answers is not in the document]
Comment (Velfi) [reply design/src/rfcs/rfc0007_timeout_config.md]: Do we have any alternative at this point?

**Review by jdisanti — APPROVED** [review]

**Pull request merged**